### ISSUE TITLE
Questions: How can I measure overfitting? How can I know which validation examples have not been successful?

### ISSUE THREAD

**Issue by marmolpen3** [huggingface/setfit#365]
Hi, thanks for sharing SetFit. 

I am using this framework to solve a specific task in my master thesis. I would like to somehow analyze the results obtained and especially to make sure that the model is not overfitting. I think by using an epoch we reduce the chances of overfitting.

I hope to hear your response. 
Thank you very much in advance.

**Comment by kgourgou** [comment]
Hello! 

I've had to think along similar lines for my work with SetFit and I think the usual data-hygiene rules apply. Are you able to split your data in parts, e.g., train/val/test, or compute metrics over cross-validations for both training and validation splits? What do the final / summary metrics look like as a function of the `num_iterations` parameter?

I also found that it helps to visualize the embeddings before and after finetunning, e.g., with TSNE, to see if the contrastive objective severely changed the structure. 

Hopefully those are somewhat useful.

**Comment by tomaarsen** [comment]
You could technically, but only through
```python
num_epochs = 5
trainer = SetFitTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    num_epochs=1,
    ...,
)

for epoch in range(5):
    trainer.train()
    train_metrics = trainer.evaluate(train_dataset)
    eval_metrics = trainer.evaluate()
    print(f"train metrics at epoch {epoch}:", train_metrics)
    print(f"eval metrics at epoch {epoch}:", eval_metrics)
```

`trainer.model_body.encode` is exactly right, well done on spotting that. You can indeed use that before and after training to visualize embeddings. In fact, that's exactly something I've done in the past. This was a visualization of my final findings:

![image](https://github.com/huggingface/setfit/assets/37621491/93e7d5ee-c26a-4980-ae94-4fe34229df22)

In short, the training accuracy went to essentially 100%, while the evaluation performance did not. These are plotted using t-SNE.

- Tom Aarsen

**Comment by marmolpen3** [comment]
How interesting, could you share the source code with the resolution or the solution of a similar example?

Thank you very much,
Elena

**Comment by kgourgou** [comment]
Hi @marmolpen3 , can you be a bit more specific?

**Comment by grofte** [comment]
@tomaarsen 

You can't do
> ```python
>    train_metrics = trainer.evaluate(train_dataset)
> ```

It's not an argument on the trainer class evaluate method. At least not on version 0.7.0

But I've added a PR for it: https://github.com/huggingface/setfit/pull/402

**Comment by grofte** [comment]
And @marmolpen3 is right - the example models here on something like sst5 are very overfitted to the training data. Probably because the sentence transformers typically output vectors of size 768. That's a lot of features.

**Comment by marmolpen3** [comment]
Sorry for not answering sooner, @kgourgou, but I'm new to NLP and I'm having a hard time understanding how it works. 

The results I get are similar to what @tomaarsen shows. I have tried with several datasets and usually the model fits too much to the training data, I guess it may be due to the contrastive type of training (I'm not sure). 

The figures I attach are my results when fitting a model with data I have labeled to solve a particular problem. I share the link to the original notebook in case it helps [notebook](https://github.com/isa-group/iContracts/blob/main/notebooks/setfit_obligations_and_rights_paraphrase_mpnet_base_v2.ipynb).

<img width="729" alt="Screenshot 2023-08-24 at 10 41 38" src="https://github.com/huggingface/setfit/assets/44238816/0b721395-f9ee-4ded-a714-81c1ea2b88f0">
<img width="1041" alt="Screenshot 2023-08-24 at 10 42 29" src="https://github.com/huggingface/setfit/assets/44238816/34557062-ce2f-4450-b09b-652d60d3f928">
<img width="1041" alt="Screenshot 2023-08-24 at 10 42 48" src="https://github.com/huggingface/setfit/assets/44238816/17241f89-290c-4601-a59e-648489d17b82">

On the other hand, thank @grofte for your contribution.

**Comment by tomaarsen** [comment]
The upcoming v1.0.0 release will include evaluation during training, allowing you to track overfitting. Furthermore, you will be able to use callbacks to e.g. plot the training or evaluation embeddings over time during training. See an example in the preliminary docs: https://moon-ci-docs.huggingface.co/docs/setfit/pr_439/en/how_to/callbacks#custom-callbacks

- Tom Aarsen
